### ISSUE TITLE
Warn if non-array key is used as array key in array creation

### ISSUE THREAD

**Issue by marcosh** [vimeo/psalm#4303]
The following code (see https://psalm.dev/r/ff5e4a33e5)

```
/**
 * @param mixed $key
 * @param mixed $value
 */
function foo($key, $value): array
{
	return [$key => $value];
}
```

breaks if `$key` is not something of `array-key` type.

Would it be possible to generate an error if a value not of `array-key` type is used as an array key?

**Comment by muglug** [comment]
This should also catch, more basically,

```php
$s = new stdClass();
$t = [$s => "hello"];
```

**Comment by orklah** [comment]
A lot of cases were handled here: https://github.com/vimeo/psalm/pull/4660

The mixed case is not covered though, I didn't know if it was supposed to play nicely with Mixed issues at level 1.

**Comment by muglug** [comment]
This is now caught
